### ISSUE TITLE
Nest one-to-one relationships

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#347]
Right now we have several one-to-one relationships in the Schema, such as 

```
post( id:"...") { 
   featuredImage { 
      id
      caption 
   } 
}
```

But often time there is edge data, that's useful in the space between the 2 connections. 

For example, we have some "featured image settings" in a metabox on our post edit screen. From a Schema perspective, the settings make sense under: 

```
post(...) { 
   featuredImage { 
       settings { 
           someField 
       } 
   } 
}
```

But that's not really easy to resolve with the current architecture where `featuredImage` returns a `MediaItem` type, and we don't want `settings` on that Type, as it's really only relevant in the context of the `post<->featuredImage` connection. . .it's "edge" data.

So, after chatting with Lee Byron at GraphQL Summit, he actually proposed one-to-one relations like this have a structure like: 

```
{
  post() {
     featuredImage {
         settings {
             someEdgeField
         }
         node { 
            id
            caption
         }
     }
  }
}
```

**Comment by CodeProKid** [comment]
Still to be resolved here: what does the registration API need to look like for this? Something like:
```
register_graphql_connection(
  [
    ...
    'one_to_one' => true,
  ]
);
```
hat flag would just prevent the pagination args from being scaffolded, and adjust the shape to not have:
`field->edges->node`, but just `field->node`

**Comment by jasonbahl** [comment]
Per @kidunot89 we should consider adding: 

- `register_graphql_edge_field()`  
- `register_graphql_connection_field()`

@kidunot89 can you expand thoughts on how you see this working?

**Comment by jasonbahl** [comment]
Some ideas here (https://wp-graphql.slack.com/archives/CDE8Y3MN3/p1580317968056400) but would . be good to get them listed out here in this issue for context.

**Comment by jasonbahl** [comment]
This is addressed in #1337 

We can now register a connection as `oneToOne => true` and it will scaffold the shape of the Schema to be `field->edge->node`

Then, we can make use of Connection Resolvers to resolve data in a one-to-one connection shape. 

Like so: 

```
$resolver = new UserConnectionResolver( $source, $args, $context, $info );
return $resolver->one_to_one()->get_connection();
```

By specifying the `$resolver->one_to_one()` it prepares the connection response to be in the shape of an edge and node, instead of a list of edges/nodes. 

Here's an example:

The `Post.author` field used to not support edge data, but is now a formal one-to-one connection. 

**Before:**
```
{
  post(id: "cG9zdDo2MTQ=") {
    title
    uri
    author { # Note, there is no space for "edge" data
      id
      name
    }
  }
}
```

**After:**

```graphql
{
  post(id: "cG9zdDo2MTQ=") {
    title
    uri
    author {
      node { # Note, there IS space for "edge" data
        id
        name
      }
    }
  }
}
```

Here's how it's registered: 

```php
register_graphql_connection( [
  'fromType'      => 'NodeWithAuthor',
  'toType'        => 'User',
  'fromFieldName' => 'author',
  'oneToOne'      => true,
  'resolve'       => function( Post $post, $args, AppContext $context, ResolveInfo $info ) {
    $resolver = new UserConnectionResolver( $post, $args, $context, $info );
    $resolver->set_query_arg( 'include', [ $post->authorDatabaseId ] );
    return $resolver->one_to_one()->get_connection();
  },
] );
```
